api/s2n.h

```diff
@@ -719,6 +719,241 @@ extern int s2n_config_set_key_log_cb(struct s2n_config *config, s2n_key_log_fn c
+ *
+ * If early data was rejected or not requested, the remaining early data size is 0.
+ * Otherwise, the remaining early data size is the maximum early data allowed by the connection,
+ * minus the early data sent or received so far.
+ *
+ * @param conn A pointer to the connection
+ * @param allowed_early_data_size A pointer which will be set to the remaining early data currently allowed by `conn`
+ * @return A POSIX error signal.
+ */
+S2N_API int s2n_connection_get_remaining_early_data_size(struct s2n_connection *conn, uint32_t *allowed_early_data_size);
+
+/**
+ * Reports the maximum size of the early data allowed by a connection.
+ *
+ * This is the maximum amount of early data that can be sent and received, assuming that early data
+ * is requested and accepted.
+ *
+ * @param conn A pointer to the connection
+ * @param max_early_data_size A pointer which will be set to the maximum early data allowed by `conn`
+ * @return A POSIX error signal.
+ */
+S2N_API int s2n_connection_get_max_early_data_size(struct s2n_connection *conn, uint32_t *max_early_data_size);
+
+/**
+ * Called by the client to begin negotiation and send early data.
+ *
+ * See https://github.com/aws/s2n-tls/blob/main/docs/USAGE-GUIDE.md#using-early-data--0rtt
+ * for usage and examples. DO NOT USE unless you have considered the security issues and
+ * implemented mitigation for anti-replay attacks.
+ *
+ * @param conn A pointer to the connection
+ * @param data A pointer to the early data to be sent
+ * @param data_len The size of the early data to send
+ * @param data_sent A pointer which will be set to the size of the early data sent
+ * @param blocked A pointer which will be set to the blocked status, as in `s2n_negotiate`.
+ * @return A POSIX error signal. The error should be handled as in `s2n_negotiate`.
+ */
+S2N_API int s2n_send_early_data(struct s2n_connection *conn, const uint8_t *data, ssize_t data_len,
+        ssize_t *data_sent, s2n_blocked_status *blocked);
+
+/**
+ * Called by the server to begin negotiation and accept any early data the client sends.
+ *
+ * See https://github.com/aws/s2n-tls/blob/main/docs/USAGE-GUIDE.md#using-early-data--0rtt
+ * for usage and examples. DO NOT USE unless you have considered the security issues and
+ * implemented mitigation for anti-replay attacks.
+ *
+ * @param conn A pointer to the connection
+ * @param data A pointer to a buffer to store the early data received
+ * @param max_data_len The size of the early data buffer
+ * @param data_received A pointer which will be set to the size of the early data received
+ * @param blocked A pointer which will be set to the blocked status, as in `s2n_negotiate`.
+ * @return A POSIX error signal. The error should be handled as in `s2n_negotiate`.
+ */
+S2N_API int s2n_recv_early_data(struct s2n_connection *conn, uint8_t *data, ssize_t max_data_len,
+        ssize_t *data_received, s2n_blocked_status *blocked);
+
+struct s2n_offered_early_data;
```

Comment:
this empty struct is confusing for me. why is it empty and how should one interpret it?

Reply:
This is a C thing. Think of it like a function declaration: we're just saying it exists, and the actual body / implementation is elsewhere. We usually call this an "opaque" structure: since the user only has the declaration, not the implementation, they can only pass around pointers to the structure and can't access any of its fields.

Reply:
ahh so this is an opaque struct. i thought any private struct was an opaque struct and i guess thinking more this is how you make things private in c